vcd/resource_vcd_vapp_vm.go

```diff
@@ -816,129 +865,172 @@ func attachDetachDisks(d *schema.ResourceData, vm govcd.VM, vdc *govcd.Vdc) erro
+	vappName := d.Get("vapp_name").(string)
+	vapp, err := vdc.GetVAppByName(vappName, false)
 
 	if err != nil {
-		return fmt.Errorf("error finding vApp: %s", err)
+		return fmt.Errorf("[VM read] error finding vApp: %s", err)
 	}
 
-	vm, err := vapp.GetVMByName(d.Get("name").(string), false)
+	identifier := d.Id()
+	if identifier == "" {
+		identifier = d.Get("name").(string)
+	}
+	if identifier == "" {
+		return fmt.Errorf("[VM read] neither name or ID were set for this VM")
+	}
+	vm, err := vapp.GetVMByNameOrId(identifier, false)
 
 	if err != nil {
-		d.SetId("")
-		return fmt.Errorf("error getting VM : %#v", err)
+		if origin == "resource" {
+			log.Printf("[DEBUG] Unable to find VM. Removing from tfstate")
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("[VM read] error getting VM : %s", err)
 	}
 
+	// org, vdc, and vapp_name are already implicitly set
 	_ = d.Set("name", vm.VM.Name)
+	_ = d.Set("description", vm.VM.Description)
+	d.SetId(vm.VM.ID)
 
+	var networkName string
+	var vappNetworkName string
+	networkNameRaw, ok := d.GetOk("network_name")
+	if ok {
+		networkName = networkNameRaw.(string)
+	}
+	vappNetworkNameRaw, ok := d.GetOk("vapp_network_name")
+	if ok {
+		vappNetworkName = vappNetworkNameRaw.(string)
+	}
 	// Read either new or deprecated networks configuration based on which are used
 	switch {
 	// TODO v3.0 remove this case block when we cleanup deprecated 'ip' and 'network_name' attributes
-	case d.Get("network_name").(string) != "" || d.Get("vapp_network_name").(string) != "":
-		ip, mac, err := deprecatedReadNetworks(d.Get("network_name").(string), d.Get("vapp_network_name").(string), *vm)
+	case networkName != "" || vappNetworkName != "":
+		ip, mac, err := deprecatedReadNetworks(networkName, vappNetworkName, *vm)
 		if err != nil {
-			return fmt.Errorf("failed reading network details: %s", err)
+			return fmt.Errorf("[ VM read] failed reading network details: %s", err)
 		}
-		d.Set("ip", ip)
-		d.Set("mac", mac)
+		_ = d.Set("ip", ip)
+		_ = d.Set("mac", mac)
 		// TODO v3.0 EO remove this case block when we cleanup deprecated 'ip' and 'network_name' attributes
-	case len(d.Get("network").([]interface{})) > 0:
+	default:
+		//case len(d.Get("network").([]interface{})) > 0:
 		networks, err := readNetworks(*vm, *vapp)
 		if err != nil {
-			return fmt.Errorf("failed reading network details: %s", err)
+			return fmt.Errorf("[ VM read] failed reading network details: %s", err)
+		}
+
+		err = d.Set("network", networks)
+		if err != nil {
+			return err
 		}
-		d.Set("network", networks)
 	}
 
-	d.Set("href", vm.VM.HREF)
-	d.Set("expose_hardware_virtualization", vm.VM.NestedHypervisorEnabled)
+	_ = d.Set("href", vm.VM.HREF)
+	_ = d.Set("expose_hardware_virtualization", vm.VM.NestedHypervisorEnabled)
+
+	cpus := 0
+	coresPerSocket := 0
+	memory := 0
+	for _, item := range vm.VM.VirtualHardwareSection.Item {
+		if item.ResourceType == 3 {
+			cpus += item.VirtualQuantity
```

Comment:
Is this being summed up from multiple resources? (the `+=` assignment)

Reply:
That's my understanding of the structure.

Reply:
I think looking at XML I saw only one structure, but this shouldn't make any harm anyway.